File: spike/presence/presence.py
#!/usr/bin/env python3
"""
User-presence detection — the harvest agent's primary control.

E2 showed contention lands in the tail: a 25 GB background load moved median
frame time 3% but p99 82%. That damage only matters while a human is watching.
At 2am it is irrelevant. So the agent should not try to be simultaneously fast
and invisible — it should detect presence and switch modes. Memory ceiling and
QoS are secondary dials applied *within* a mode.

Every signal here is read through IOKit or power management rather than through
AppKit or the window server, because a LaunchDaemon in session 0 has no GUI
session. NSWorkspace and CGEventSource would be richer but are unavailable
there; whether even these survive session 0 is what E1 measures.

Two properties the design turns on:

1. Asymmetric time constants. Yield must be near-instant, because the moment a
   user returns is the moment they would notice. Resume must be slow and
   conservative, because a false "they're gone" costs a model load and an
   immediate preemption. Hysteresis is not a refinement here, it is the point.

2. Idle time alone is wrong. Someone on a video call, watching playback, or
   waiting on a render produces no HID events for many minutes while being
   entirely present. Power assertions catch exactly that case, which is why
   PASSIVE exists as a distinct state rather than collapsing into IDLE.
"""

import argparse
import json
import subprocess
import time

# How often to sample. E4 measured memory release at ~20ms and model reload at
# 1-3s, which makes this interval the dominant term in end-to-end yield latency:
# polling every 2s means up to 2s of work continues after a user touches the
# keyboard, ~100x the cost of the release it triggers. Sampling is cheap (ioreg
# and pmset reads), so poll fast. Tune this, not the release path.
POLL_INTERVAL_ACTIVE = 0.5   # user present or recently present
POLL_INTERVAL_IDLE = 5.0     # confirmed idle; nothing to interrupt quickly

# Seconds of no HID input before the machine stops counting as actively used.
ACTIVE_IDLE_THRESHOLD = 90
# Sustained idle required before promoting to a more permissive state. Long on
# purpose: the cost of resuming too early is a wasted model load.
IDLE_PROMOTE_SECONDS = 300

# Ordered least to most permissive. Promotion needs hysteresis; demotion is
# immediate.
STATES = ["ACTIVE", "PASSIVE", "IDLE", "LOCKED", "ABSENT"]

# What the agent may do in each state.
#
# The GPU/ANE split is the most important column here. E5 measured a saturating
# ANE workload as statistically indistinguishable from no load at all (p95
# -16%, inside a 36% baseline noise floor) while equivalent GPU work cost
# +59-100%. The ANE is separate silicon and an artist's viewport does not touch
# it, so ANE work stays permitted in states where GPU work is forbidden.
#
# mem_frac is NOT a politeness dial and must not be treated as one: E5 saw a
# 4 GB load — a small fraction of a 64 GB machine — degrade viewport p95 by
# 100%. Footprint governs what fits, not how much a user is disturbed.
# Throttling disturbance requires limiting compute rate, which this table does
# not yet express. See e5_ane/FINDINGS.md.
POLICY = {
    "ACTIVE":  {"gpu": False, "ane": False, "qos": None,         "mem_frac": 0.00},
    # ANE stays on through PASSIVE and even ACTIVE-adjacent states on the E5
    # result; GPU does not.
    "PASSIVE": {"gpu": False, "ane": True,  "qos": "background", "mem_frac": 0.15},
    "IDLE":    {"gpu": True,  "ane": True,  "qos": "background", "mem_frac": 0.35},
    "LOCKED":  {"gpu": True,  "ane": True,  "qos": "standard",   "mem_frac": 0.70},
    "ABSENT":  {"gpu": True,  "ane": True,  "qos": "standard",   "mem_frac": 0.85},
}


def _sh(cmd):
    try:
        return subprocess.run(cmd, shell=True, capture_output=True,
                              text=True, timeout=10).stdout
    except Exception:
        return ""


def hid_idle_seconds():
    """Seconds since the last keyboard/mouse/trackpad event.

    Read from IOHIDSystem rather than CGEventSourceSecondsSinceLastEventType so
    it does not need a GUI session. Value is nanoseconds.
    """
    out = _sh("ioreg -c IOHIDSystem 2>/dev/null | grep -m1 HIDIdleTime")
    try:
        return int(out.split("=")[-1].strip()) / 1e9
    except (ValueError, IndexError):
        return None


def screen_locked():
    out = _sh("ioreg -n Root -d1 -a 2>/dev/null | "
              "plutil -extract IOConsoleLocked raw - 2>/dev/null").strip()
    return {"true": True, "false": False}.get(out)


def console_user():
    user = _sh("stat -f%Su /dev/console").strip()
    # loginwindow owns the console when nobody is logged in.
    return None if user in ("", "root", "_windowserver", "loginwindow") else user


def on_ac_power():
    return "AC Power" in _sh("pmset -g batt")


# Background services that hold sleep assertions indefinitely with no human
# involved. Without this list sharingd's permanent "Handoff" assertion pins a
# machine in PASSIVE forever and it never harvests at all.
SYSTEM_ASSERTION_PROCS = {
    "powerd", "sharingd", "backupd", "mds", "mds_stores", "mDNSResponder",
    "softwareupdated", "nsurlsessiond", "cloudd", "bird", "AppleIDAuthAgent",
    "UpdateBrainService", "corespeechd", "photoanalysisd", "AMPDeviceDiscoveryAgent",
}


def _parse_assertions():
    """Split sleep assertions by what they actually imply.

    These answer two different questions and conflating them is wrong:

      PreventUserIdleDisplaySleep — something insists the *display* stay on.
        Video calls, playback, presentations. Strong evidence a human is
        looking at the screen right now.

      PreventUserIdleSystemSleep — something insists the *machine* keep
        running. Renders and long jobs hold this, but so do a dozen background
        daemons permanently. Evidence the machine is busy, not that anyone is
        present.

    Only the display class implies presence. The system class implies
    contention, which is a separate reason to back off.
    """
    display, system = [], []
    for line in _sh("pmset -g assertions 2>/dev/null").splitlines():
        line = line.strip()
        if not line.startswith("pid "):
            continue
        proc = line.split("(", 1)[-1].split(")", 1)[0] if "(" in line else ""
        if proc in SYSTEM_ASSERTION_PROCS:
            continue
        name = line.split("named:")[-1].strip().strip('"') if "named:" in line else line
        label = f"{proc}: {name}"[:60]
        if "PreventUserIdleDisplaySleep" in line:
            display.append(label)
        elif "PreventUserIdleSystemSleep" in line:
            system.append(label)
    return display, system


def thermal_ok():
    out = _sh("pmset -g therm")
    return not any(w in out for w in ("CPU_Speed_Limit", "warning level 1",
                                      "warning level 2", "warning level 3"))


def read_signals():
    idle = hid_idle_seconds()
    display_a, busy_a = _parse_assertions()
    return {
        "timestamp": time.time(),
        "hid_idle_s": round(idle, 1) if idle is not None else None,
        "screen_locked": screen_locked(),
        "console_user": console_user(),
        "on_ac_power": on_ac_power(),
        "display_assertions": display_a,   # implies a human is watching
        "busy_assertions": busy_a,          # implies the machine is busy
        "thermal_ok": thermal_ok(),
    }


def classify(sig):
    """Map raw signals to a presence state.

    Order matters: the most restrictive interpretation that fits wins. A missing
    signal is treated as "user present", because the failure mode of guessing
    absent is degrading someone's machine, and that ends the program.
    """
    if sig["console_user"] is None:
        return "ABSENT"
    if sig["screen_locked"] is True:
        return "LOCKED"

    idle = sig["hid_idle_s"]
    if idle is None:
        return "ACTIVE"  # cannot tell; assume the worst
    if idle < ACTIVE_IDLE_THRESHOLD:
        return "ACTIVE"
    if sig["display_assertions"]:
        # Something is holding the display awake: a call, playback, a
        # presentation. No keystrokes, but a human is almost certainly looking.
        return "PASSIVE"
    return "IDLE"


def effective_policy(state, sig):
    """Apply hard gates that override the state's policy entirely."""
    policy = dict(POLICY[state])
    reasons = []
    if not sig["on_ac_power"]:
        policy.update(gpu=False, ane=False, mem_frac=0.0)
        reasons.append("on battery")
    if not sig["thermal_ok"]:
        policy.update(gpu=False, mem_frac=0.0)
        reasons.append("thermal pressure")
    if sig["busy_assertions"]:
        # A render or long job is running even though nobody is typing. Not a
        # presence signal, but competing with it wastes both workloads.
        policy["mem_frac"] = min(policy["mem_frac"], 0.15)
        policy["gpu"] = False
        reasons.append(f"machine busy ({len(sig['busy_assertions'])} assertion/s)")
    policy["blocked_by"] = reasons
    return policy


class PresenceMonitor:
    """Applies hysteresis so the agent does not flap between states.

    Demotion (toward ACTIVE) is immediate — a returning user must be respected
    on the very first sample. Promotion (toward ABSENT) requires the condition
    to hold for IDLE_PROMOTE_SECONDS, so a momentary lull does not trigger an
    expensive model load that is about to be thrown away.
    """

    def __init__(self, promote_after=IDLE_PROMOTE_SECONDS):
        self.promote_after = promote_after
        self.state = "ACTIVE"
        self._candidate = None
        self._candidate_since = None

    def update(self, sig=None, now=None):
        sig = sig or read_signals()
        now = now if now is not None else sig["timestamp"]
        observed = classify(sig)

        if STATES.index(observed) <= STATES.index(self.state):
            # More restrictive (or unchanged): apply immediately.
            self.state = observed
            self._candidate = None
        else:
            if observed != self._candidate:
                self._candidate = observed
                self._candidate_since = now
            elif now - self._candidate_since >= self.promote_after:
                self.state = observed
                self._candidate = None

        return {
            "state": self.state,
            "observed": observed,
            "pending": self._candidate,
            "pending_for_s": (round(now - self._candidate_since, 1)
                              if self._candidate else None),
            "policy": effective_policy(self.state, sig),
            "signals": sig,
        }


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--watch", action="store_true", help="poll continuously")
    ap.add_argument("--interval", type=float, default=POLL_INTERVAL_ACTIVE)
    ap.add_argument("--json", action="store_true", help="one-shot JSON")
    args = ap.parse_args()

    monitor = PresenceMonitor()

    if args.json:
        print(json.dumps(monitor.update(), indent=2))
        return 0

    if not args.watch:
        sig = read_signals()
        print(json.dumps({"state": classify(sig), "signals": sig}, indent=2))
        return 0

    print(f"{'state':<9} {'observed':<9} {'idle':>7} {'lock':>6} {'ac':>4} "
          f"{'gpu':>4} {'qos':>11} {'mem':>5}  assertions")
    print("-" * 82)
    while True:
        r = monitor.update()
        s, p = r["signals"], r["policy"]
        print(f"{r['state']:<9} {r['observed']:<9} "
              f"{(s['hid_idle_s'] if s['hid_idle_s'] is not None else -1):>7.1f} "
              f"{str(s['screen_locked']):>6} {str(s['on_ac_power']):>4} "
              f"{str(p['gpu']):>4} {str(p['qos']):>11} {p['mem_frac']:>5.2f}  "
              f"{','.join(s['display_assertions'] + s['busy_assertions'])[:34]}",
              flush=True)
        time.sleep(args.interval)


if __name__ == "__main__":
    raise SystemExit(main())

File: agent/Sources/DaiAgent/Presence.swift
import Foundation

/// User-presence detection: the harvest agent's primary control.
///
/// E2 showed contention lands in the tail. A 25 GB background load moved median
/// frame time 3% but p99 82%, and that damage only matters while a human is
/// watching. At 2am it is irrelevant. So the agent does not try to be
/// simultaneously fast and invisible; it detects presence and switches modes.
/// Memory ceiling and QoS are secondary dials applied *within* a mode.
///
/// Everything here is pure. Signal collection lives behind ``SignalSource`` so
/// tests inject fixtures: every policy bug found in the Python agent reproduced
/// from a recorded signal dictionary alone, with no hardware involved, and that
/// property is worth protecting.

public enum PresenceState: String, Sendable, CaseIterable, Comparable {
    /// Ordered least to most permissive. Promotion needs hysteresis; demotion is
    /// immediate.
    case active = "ACTIVE"
    case passive = "PASSIVE"
    case idle = "IDLE"
    case locked = "LOCKED"
    case absent = "ABSENT"

    var rank: Int { Self.allCases.firstIndex(of: self)! }
    public static func < (a: PresenceState, b: PresenceState) -> Bool { a.rank < b.rank }
}

public enum WorkKind: String, Sendable, CaseIterable {
    case embed, generate, render

    /// `embed` runs on the ANE. The other two are GPU work and are confined to
    /// states where nobody is at the machine.
    public var isGPU: Bool { self != .embed }

    /// Whether anything in this build can actually run the kind.
    ///
    /// All three now are. Kept as a value rather than deleted because it is what
    /// a fourth kind will need on the day it is declared and before it works,
    /// and because the alternative - inferring it from a `throw` buried in the
    /// batch loop - is what let `render` be advertised for months while
    /// answering every unit with an error.
    public var isImplemented: Bool { true }
}

public enum QoS: String, Sendable { case background, standard }

public struct StatePolicy: Sendable, Equatable {
    public var gpu: Bool
    public var ane: Bool
    public var qos: QoS
    public var dutyMax: Double
    public var memFrac: Double
    /// Largest completion this state permits, in tokens.
    ///
    /// Batch work yields between items and hands back the remainder. A single
    /// interactive request has no such seam, so bounding its length bounds how
    /// long a returning user waits for their machine back.
    public var maxCompletionTokens: Int
    public var blockedBy: [String] = []
}

/// Seconds of no HID input before the machine stops counting as actively used.
public let activeIdleThreshold: TimeInterval = 90

/// Sustained idle required before adopting a more permissive state. Long on
/// purpose: E4 showed a false "they are gone" costs a model load and an
/// immediate preemption.
public let idlePromoteSeconds: TimeInterval = 300

/// How often to sample. E4 measured memory release at ~20ms and model reload at
/// 1-3s, which makes this interval the dominant term in end-to-end yield
/// latency. Sampling is cheap, so poll fast; tune this rather than the release
/// path.
public let pollIntervalActive: TimeInterval = 0.5
public let pollIntervalIdle: TimeInterval = 5.0

/// What the agent may do in each state. These are E2 and E5 measurements, not
/// estimates, and the control plane serves the same table.
///
/// GPU work is forbidden wherever a user is logged in: E2 swept QoS against duty
/// cycle and found every configuration perceptible, the gentlest tested
/// (background QoS at 25% duty) still costing 46% of viewport p95.
///
/// ANE work is permitted everywhere. E5 measured a saturating ANE workload as
/// indistinguishable from no load, which makes it the only daytime option and
/// the only thing three of five states allow at all.
///
/// `memFrac` is **not** a politeness dial. E2 measured a 32 GB load disturbing a
/// viewport less than an 8 GB one at identical duty. Footprint governs what
/// fits; occupancy governs disturbance.
public let defaultPolicy: [PresenceState: StatePolicy] = [
    .active:  StatePolicy(gpu: false, ane: true, qos: .background, dutyMax: 0, memFrac: 0, maxCompletionTokens: 256),
    .passive: StatePolicy(gpu: false, ane: true, qos: .background, dutyMax: 0, memFrac: 0.15, maxCompletionTokens: 256),
    .idle:    StatePolicy(gpu: false, ane: true, qos: .background, dutyMax: 0, memFrac: 0.35, maxCompletionTokens: 256),
    .locked:  StatePolicy(gpu: true, ane: true, qos: .standard, dutyMax: 1, memFrac: 0.70, maxCompletionTokens: 2048),
    .absent:  StatePolicy(gpu: true, ane: true, qos: .standard, dutyMax: 1, memFrac: 0.85, maxCompletionTokens: 4096),
]

public struct Signals: Sendable, Equatable {
    /// Seconds since the last keyboard, mouse or trackpad event. `nil` when the
    /// signal could not be read at all.
    public var hidIdleSeconds: TimeInterval?
    public var screenLocked: Bool?
    public var consoleUser: String?
    public var onACPower: Bool
    /// Something insists the *display* stay on: a call, playback, a
    /// presentation. Strong evidence a human is looking at the screen.
    public var displayAssertions: [String]
    /// Something insists the *machine* keep running. Renders hold this, but so
    /// do a dozen background daemons, so it implies the machine is busy rather
    /// than that anyone is present.
    public var busyAssertions: [String]
    public var thermalOK: Bool

    public init(hidIdleSeconds: TimeInterval? = 600, screenLocked: Bool? = false,
                consoleUser: String? = "user", onACPower: Bool = true,
                displayAssertions: [String] = [], busyAssertions: [String] = [],
                thermalOK: Bool = true) {
        self.hidIdleSeconds = hidIdleSeconds
        self.screenLocked = screenLocked
        self.consoleUser = consoleUser
        self.onACPower = onACPower
        self.displayAssertions = displayAssertions
        self.busyAssertions = busyAssertions
        self.thermalOK = thermalOK
    }
}

/// Map raw signals to a presence state.
///
/// Order matters: the most restrictive interpretation that fits wins. A missing
/// signal is treated as "user present", because the failure mode of guessing
/// absent is degrading someone's machine, and that ends the programme.
public func classify(_ s: Signals) -> PresenceState {
    if s.consoleUser == nil { return .absent }
    if s.screenLocked == true { return .locked }
    guard let idle = s.hidIdleSeconds else { return .active }  // cannot tell: assume the worst
    if idle < activeIdleThreshold { return .active }
    // Present but not typing. Someone on a call or watching playback produces no
    // HID events for many minutes while being entirely there, which is why this
    // is a distinct state rather than folded into idle.
    if !s.displayAssertions.isEmpty { return .passive }
    return .idle
}

/// Apply hard gates that override the state's policy entirely.
public func effectivePolicy(_ state: PresenceState, _ s: Signals) -> StatePolicy {
    var p = defaultPolicy[state]!
    var reasons: [String] = []

    if !s.onACPower {
        // The gate that would have saved a laptop which drained its battery
        // running a worker with no presence logic at all.
        p.gpu = false; p.ane = false; p.dutyMax = 0; p.memFrac = 0
        reasons.append("on battery")
    }
    if !s.thermalOK {
        p.gpu = false; p.dutyMax = 0; p.memFrac = 0
        reasons.append("thermal pressure")
    }

    // `busyAssertions` is deliberately NOT a gate. PreventUserIdleSystemSleep
    // means only "do not sleep": Safari, coreaudiod, downloads and caffeinate
    // hold it more or less permanently, so gating on it blocked harvesting
    // entirely on a normally-used machine. It is surfaced for observability and
    // nothing else.

    p.blockedBy = reasons
    return p
}

/// Work kinds this state permits. The control plane applies the same rule, and
/// the stricter of the two wins.
///
/// A statement about policy, not about this machine. It answers "would a node in
/// this state be allowed to do it", which is why `render` appears: rendering is
/// GPU work and LOCKED and ABSENT permit GPU work. Whether any node could
/// actually run it is a different question, answered by ``runnableKinds(_:hasGPU:hasANE:)``.
public func permittedKinds(_ state: PresenceState, policy: StatePolicy? = nil) -> [WorkKind] {
    permittedKinds(policy: policy ?? defaultPolicy[state]!)
}

public func permittedKinds(policy p: StatePolicy) -> [WorkKind] {
    var kinds: [WorkKind] = []
    if p.ane { kinds.append(.embed) }
    if p.gpu && p.dutyMax > 0 { kinds.append(contentsOf: [.generate, .render]) }
    return kinds
}

/// What this machine will actually attempt: what policy permits, narrowed to
/// kinds that exist and have a runtime here.
///
/// The distinction is not pedantry. Both numbers get reported to a person
/// diagnosing a machine that is doing nothing, and they have entirely different
/// causes: "policy forbids it" is answered by waiting or by unplugging a
/// monitor, "there is no runtime for it" is answered by installing one. A single
/// line conflating the two sent a reader looking at presence when the machine
/// had no Metal toolchain.
///
/// This is also what the node offers the scheduler, so it is never handed work
/// it must immediately give back.
public func runnableKinds(_ p: StatePolicy, hasGPU: Bool, hasANE: Bool,
                          hasRenderer: Bool = false) -> [WorkKind] {
    permittedKinds(policy: p).filter { kind in
        guard kind.isImplemented else { return false }
        switch kind {
        case .embed: return hasANE
        case .generate: return hasGPU
        // A renderer, not a GPU runtime. A machine can have Blender and no MLX
        // or the reverse, and treating them as one capability would have a node
        // with a Metal toolchain and no renderer offered every render unit in
        // the queue.
        case .render: return hasRenderer
        }
    }
}

/// Applies hysteresis so the agent does not flap between states.
///
/// Demotion toward `.active` is immediate: a returning user must be respected on
/// the very first sample. Promotion requires the condition to hold for
/// `promoteAfter`, because a momentary lull should not trigger an expensive
/// model load that is about to be thrown away. The asymmetry is the design, not
/// a refinement.
public final class PresenceMonitor {
    public private(set) var state: PresenceState = .active
    private var candidate: PresenceState?
    private var candidateSince: TimeInterval = 0
    private let promoteAfter: TimeInterval

    public init(promoteAfter: TimeInterval = idlePromoteSeconds) {
        self.promoteAfter = promoteAfter
    }

    public struct Reading: Sendable {
        public let state: PresenceState
        public let observed: PresenceState
        public let policy: StatePolicy
        public let signals: Signals
    }

    @discardableResult
    public func update(_ signals: Signals, now: TimeInterval) -> Reading {
        let observed = classify(signals)

        if observed <= state {
            state = observed          // more restrictive, or unchanged: apply now
            candidate = nil
        } else if observed != candidate {
            candidate = observed
            candidateSince = now
        } else if now - candidateSince >= promoteAfter {
            state = observed
            candidate = nil
        }

        return Reading(state: state, observed: observed,
                       policy: effectivePolicy(state, signals), signals: signals)
    }

    /// Sample fast when a user could plausibly be about to return; slowly when
    /// nobody can see the screen.
    public var pollInterval: TimeInterval {
        (state == .locked || state == .absent) ? pollIntervalIdle : pollIntervalActive
    }
}
